fdbclient/include/fdbclient/BlobCipher.h

```diff
@@ -181,6 +192,261 @@ struct hash<BlobCipherDetails> {
+	  : cipherTextDetails(textDetails) {
+		ASSERT_EQ(ivLen, AES_256_IV_LENGTH);
+		memcpy(&iv[0], ivBuf, ivLen);
+	}
+
+	bool operator==(const AesCtrNoAuthV1& o) const {
+		return cipherTextDetails == o.cipherTextDetails && memcmp(&iv[0], &o.iv[0], AES_256_IV_LENGTH) == 0;
+	}
+
+	static Standalone<StringRef> toStringRef(const AesCtrNoAuthV1& algoHeader, Arena& arena) {
+		BinaryWriter wr(AssumeVersion(ProtocolVersion::withEncryptionAtRest()));
+		wr.serializeBytes(&algoHeader, sizeof(AesCtrNoAuthV1));
+		return wr.toValue(arena);
+	}
+
+	template <class Ar>
+	void serialize(Ar& ar) {
+		serializer(ar, cipherTextDetails);
+		ar.serializeBytes(iv, AES_256_IV_LENGTH);
+	}
+};
+
+struct BlobCipherEncryptHeaderRef {
+	// Serializable fields
+
+	// HeaderFlags version tracker
+	uint16_t flagsVersion;
+	// Encryption algorithm header version tracker
+	uint16_t algoHeaderVersion;
+
+	// The on-disk format doesn't store std::variant, currently "serializer" doesn't support std::variant, the
+	// (de)serialization code serializes the relevant BlobCipherEncryptHeader and AlgoHeader structs for a given
+	// 'flagVersion' and 'algoHeaderVersion'. Refer BlobCipherEncryptHeaderRef::serialize() for more details.
+	std::variant<BlobCipherEncryptHeaderFlagsV1> flags;
+	std::variant<AesCtrNoAuthV1, AesCtrWithAuthV1<AUTH_TOKEN_HMAC_SHA_SIZE>, AesCtrWithAuthV1<AUTH_TOKEN_AES_CMAC_SIZE>>
+	    algoHeader;
+
+	BlobCipherEncryptHeaderRef()
+	  : flagsVersion(INVALID_ENCRYPT_HEADERS_FLAG_VERSION),
+	    algoHeaderVersion(INVALID_ENCRYPT_HEADER_ALGO_HEADER_VERSION) {}
+	BlobCipherEncryptHeaderRef(const BlobCipherEncryptHeaderRef& src)
+	  : flagsVersion(src.flagsVersion), algoHeaderVersion(src.algoHeaderVersion), flags(src.flags),
+	    algoHeader(src.algoHeader) {}
+
+	static BlobCipherEncryptHeaderRef fromStringRef(const StringRef& header) {
+		return BinaryReader::fromStringRef<BlobCipherEncryptHeaderRef>(
+		    header, AssumeVersion(ProtocolVersion::withEncryptionAtRest()));
+	}
+
+	static Standalone<StringRef> toStringRef(const BlobCipherEncryptHeaderRef& headerRef) {
+		return BinaryWriter::toValue(headerRef, AssumeVersion(ProtocolVersion::withEncryptionAtRest()));
+	}
+
+	// Routine computes EncryptHeaderSize based on input params.
+	static uint32_t getHeaderSize(const int flagVersion,
+	                              const int algoHeaderVersion,
+	                              const EncryptCipherMode cipherMode,
+	                              const EncryptAuthTokenMode authMode,
+	                              const EncryptAuthTokenAlgo authAlgo);
+
+	template <class Ar>
+	void serialize(Ar& ar) {
+		// TODO: once std::variant native (de)serialization support is added, the method would transform to much shorter
+		// implementaion
+		uint8_t encryptMode;
+		EncryptAuthTokenMode authMode;
+		EncryptAuthTokenAlgo authAlgo;
+
+		serializer(ar, flagsVersion, algoHeaderVersion);
+		if (ar.isSerializing) {
+			if (flagsVersion != 1) {
+				throw not_implemented();
```

Comment:
nit: adding a trace log?

Reply:
Will do it in my subsequent review (CI is green, so feel like merging the change and unblocking followup changes).
Thanks